contrib/Intel/JenkinsBuildScripts/tests.py

```diff
@@ -0,0 +1,311 @@
+        self.libfab_installpath = "{}/{}/{}".format(cisite_config.install_dir, self.branchname, self.buildno) 
+        self.env = [("FI_VERBS_MR_CACHE_ENABLE", "1"),("FI_VERBS_INLINE_SIZE", "256")] if self.core_prov == "verbs" else []
+                     
+
+    def get_exclude_file(self):
+        path = self.libfab_installpath
+        efile_path = "{}/share/fabtests/test_configs".format(path)
+
+        prov = self.util_prov if self.util_prov else self.core_prov
+        efile_old = "{path}/{prov}/{prov}.exclude".format(path=efile_path, prov=prov)
+        if self.util_prov:
+            efile = "{path}/{util_prov}/{core_prov}/exclude".format(
+            path=efile_path, util_prov=self.util_prov, core_prov=self.core_prov)
+        else:
+            efile = "{path}/{prov}/exclude".format(path=efile_path, prov=self.core_prov)
+           
+        if os.path.isfile(efile):
+            return efile
+        elif os.path.isfile(efile_old):
+            return efile_old
+        else:
+            print("Exclude file: {} not found!".format(efile))
+            return None  
+
+class fabtest(test):
+    
+    def __init__(self, branchname, buildno, testname, core_prov, fabric, hosts, util_prov=None):
+        super().__init__(branchname, buildno, testname, core_prov, fabric, hosts, util_prov)
+        self.fabtestpath = "{}/bin".format(self.libfab_installpath) 
+   
+    @property    
+    def cmd(self):    
+        #return "./runfabtests.sh "
+        return "{}/runfabtests.sh ".format(self.fabtestpath)
+     
+    @property
+    def options(self):
+        opts = "-vvv -p {} -S ".format(self.fabtestpath)
+        if (self.core_prov == "verbs" and self.nw_interface):
+            opts = "{} -s {} ".format(opts, common.get_node_name(self.server, self.nw_interface)) # include common.py
+            opts = "{} -c {} ".format(opts, common.get_node_name(self.client, self.nw_interface)) # from common.py
+        if (self.core_prov == "shm"):
+            opts = "{} -s {} ".format(opts, self.server)
+            opts = "{} -c {} ".format(opts, self.client)
+            opts += "-N "
+            
+        if not re.match(".*sockets|udp|tcp.*", self.core_prov):
+            opts = "{} -t all ".format(opts)
+
+        efile = self.get_exclude_file()
+        if efile:
+            opts = "{} -R ".format(opts)
+            opts = "{} -f {} ".format(opts, efile)  
+        
+        for key,val in self.env:
+            opts = "{options} -E {key}={value} ".format(options = opts, key=key, value=val)
+    
+        if self.util_prov:
+            opts = "{options} {core};{util} ".format(options=opts, core=self.core_prov, util=self.util_prov)
+        else:
+            opts = "{options} {core} ".format(options=opts,core=self.core_prov)
+        
+        if (self.core_prov == "shm"):
+            opts += "{} {} ".format(self.client, self.server)
+        else:
+            opts += "{} {} ".format(self.server, self.client)
```

Comment:
Keep opts as a list of strings and do a join when returning it. The last time I checked that's the recommended method as it's faster (especially when multiple string formatting / concatenations are involved).

Reply:
okay, I didn't know about the speed up. I initially created opts as a list of strings essentially because subprocess requires a list. However, to format it properly as a list, I moved to using shlex, which expects string input. should I change this back, to list everywhere in the code, only to convert back into string before passing it in to shlex? Any idea if the speedup is significant to change it everywhere in the code?

Reply:
You can convert the list to string just before passing it to shlex. Not sure how much is the speedup. I remember there was some stackoverflow answer that measured it. Anyway, it's not a big concern. A followup PR can fix it but if you want to change here you can go ahead and do it.